rust/lance/src/dataset.rs

```diff
@@ -3922,4 +3925,44 @@ mod tests {
         let row_count = batches.iter().map(|batch| batch.num_rows()).sum::<usize>();
         assert_eq!(row_count, 1900);
     }
+
+    #[tokio::test]
+    async fn test_bfloat16_roundtrip() -> Result<()> {
+        let inner_field = Arc::new(
+            Field::new("item", DataType::FixedSizeBinary(2), true).with_metadata(
+                [
+                    ("ARROW:extension:name".into(), "lance.bfloat16".into()),
+                    ("ARROW:extension:metadata".into(), "".into()),
+                ]
+                .into(),
+            ),
+        );
+        let schema = Arc::new(ArrowSchema::new(vec![Field::new(
+            "fsl",
+            DataType::FixedSizeList(inner_field.clone(), 2),
+            false,
+        )]));
+
+        let values = bfloat16::BFloat16Array::from_iter_values(
+            (0..6).map(|i| i as f32).map(half::bf16::from_f32),
+        );
+        let vectors = FixedSizeListArray::new(inner_field, 2, Arc::new(values.into_inner()), None);
+
+        let batch = RecordBatch::try_new(schema.clone(), vec![Arc::new(vectors)]).unwrap();
+
+        let test_dir = tempdir().unwrap();
+        let test_uri = test_dir.path().to_str().unwrap();
+
+        let dataset = Dataset::write(
+            RecordBatchIterator::new(vec![Ok(batch.clone())], schema.clone()),
+            test_uri,
+            None,
+        )
+        .await?;
+
+        let data = dataset.scan().try_into_batch().await?;
```

Comment:
When scanning the data, could you add an example of how to check BF16 data type explicitly in the schema? Gonna use it in the indexing code.

Reply:
Lance schema or Arrow schema?

Reply:
Added a `TryFrom<&Field>` implementation that handles this: https://github.com/lancedb/lance/pull/1810/files#diff-f9807f8eeeb4c65d48f9f292b42103928e319376ec1d2ed1dfadec47b1f5022e